### PR TITLE
Determine Platform and Framework if not specified

### PR DESCRIPTION
**Pull request by smadala**
- Making platform and framework switches optional.
- On conflict platform/framework values, Chose default values.

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs [hunk 1]
```diff
@@ -84,6 +84,10 @@ public void ProcessRequests(ITestHostManagerFactory testHostManagerFactory)
             do
             {
                 var message = this.communicationManager.ReceiveMessage();
+                if (EqtTrace.IsInfoEnabled)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs]: You mean verbose? Don't think printing JSON protocol messages are informational :)

File: src/Microsoft.TestPlatform.ObjectModel/Framework.cs [hunk 2]
```diff
@@ -83,5 +89,22 @@ public override string ToString()
         {
             return this.Name;
         }
+
+        public override bool Equals(Object obj)
+        {
+            Framework other = obj as Framework;
+            if (other == null) return false;
```
Comment (codito) [inline src/Microsoft.TestPlatform.ObjectModel/Framework.cs]: nit: please use blocks for conditional statements.

File: src/Microsoft.TestPlatform.ObjectModel/Framework.cs [hunk 3]
```diff
@@ -83,5 +89,22 @@ public override string ToString()
...
+            bool equal = string.Equals(this.Name, other.Name);
```
Comment (codito) [inline src/Microsoft.TestPlatform.ObjectModel/Framework.cs]: Culture invariant? Case sensitive?

File: src/vstest.console/CommandLine/AssemblyHelper.cs [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+    using System.Reflection;
+    using System.Reflection.Metadata;
+    using System.Reflection.PortableExecutable;
+    using System.Runtime.Versioning;
+    using System.Text;
+    using ObjectModel;
+
+    internal class AssemblyHelper : IAssemblyHelper
+    {
+        private static AssemblyHelper _instance;
```
Comment (codito) [inline src/vstest.console/CommandLine/AssemblyHelper.cs]: nit: this repo doesn't use `_` for private fields.

File: src/vstest.console/CommandLine/AssemblyHelper.cs [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+        /// </summary>
+        internal static AssemblyHelper Instance => _instance ?? (_instance = new AssemblyHelper());
+
+        /// <inheritdoc />
+        public FrameworkName GetFrameWork(string filePath)
+        {
+            FrameworkName frameworkName = new FrameworkName(Framework.DefaultFramework.Name);
+            try
+            {
+                using (var assemblyStream = File.Open(filePath, FileMode.Open))
```
Comment (codito) [inline src/vstest.console/CommandLine/AssemblyHelper.cs]: Is it interesting to unit test this logic?

File: src/vstest.console/CommandLine/AssemblyHelper.cs [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+            if (EqtTrace.IsInfoEnabled)
+            {
+                EqtTrace.Info("GetArchitecture: determined atchitecture:{0} info for assembly: {1}", archType,
+                    assemblyPath);
+            }
+
+            return archType;
+        }
+        private static FrameworkName GetFrameworkNameFromAssemblyMetadata(FileStream assemblyStream,
```
Comment (codito) [inline src/vstest.console/CommandLine/AssemblyHelper.cs]: nit: new line between method declarations

File: src/vstest.console/CommandLine/AssemblyHelper.cs [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+            FrameworkName frameworkName)
```
Comment (codito) [inline src/vstest.console/CommandLine/AssemblyHelper.cs]: nit: if one arg is a new line, all args should be in new lines

File: src/vstest.console/CommandLine/Interfaces/IAssemblyHelper.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace Microsoft.VisualStudio.TestPlatform.CommandLine.Processors
+{
+    using System.Runtime.Versioning;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+
+    /// <summary>
+    /// Defines interface for assembly properties.
+    /// </summary>
+    internal interface IAssemblyHelper
```
Comment (codito) [inline src/vstest.console/CommandLine/Interfaces/IAssemblyHelper.cs]: Not sure if this logic belongs here. May be an entrypoint should not know how to read PE files.

File: src/vstest.console/CommandLine/Interfaces/IInferHelper.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace Microsoft.VisualStudio.TestPlatform.CommandLine.Processors
+{
+    using System.Collections.Generic;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+
+    /// <summary>
+    /// Defines interface for infering run configuration values.
+    /// </summary>
+    internal interface IInferHelper
```
Comment (codito) [inline src/vstest.console/CommandLine/Interfaces/IInferHelper.cs]: I think this interface doesn't add any value. There will always be one implementation, and that seems to be a singleton.

File: src/vstest.console/CommandLine/Interfaces/IInferHelper.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        /// <summary>
+        /// Determines Framework from sources.
+        /// </summary>
+        Framework AutoDetectFramework(List<string> sources);
```
Comment (codito) [inline src/vstest.console/CommandLine/Interfaces/IInferHelper.cs]: May be this should be along side other runsettings services. There's also a InferRunSettingsHelper, how do we differentiate behavior of these units? And there's the xmlrunsettingsutilities, runsettingsmanager and runsettingsprovider too.

Comment (codito) [reply src/vstest.console/CommandLine/Interfaces/IInferHelper.cs]: How do we make runsettings management cohesive?

File: src/vstest.console/CommandLine/AssemblyHelper.cs (hunk 7 above)
Comment (chakradhar2) [inline src/vstest.console/CommandLine/AssemblyHelper.cs]: Can we make frameworkName out param or local variable?

File: src/vstest.console/Processors/RunTestsArgumentProcessor.cs [hunk 11]
```diff
@@ -51,7 +51,7 @@ public Lazy<IArgumentExecutor> Executor
                         CommandLineOptions.Instance,
                         RunSettingsManager.Instance,
                         TestRequestManager.Instance,
-                        ConsoleOutput.Instance));
+                        ConsoleOutput.Instance, InferHelper.Instance));
```
Comment (chakradhar2) [inline src/vstest.console/Processors/RunTestsArgumentProcessor.cs]: nit: add new param to new line.

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs [hunk 12]
```diff
@@ -360,5 +387,33 @@ private bool RunTests(TestRunCriteria testRunCriteria, ITestRunEventsRegistrar t
...
+
+            return required;
+        }
+
+        private bool IsAutoPlatformDetectRequired(XPathNavigator navigator)
+        {
+            bool required = false;
+            bool isValidPlatform =
+                InferRunSettingsHelper.TryGetPlatformXml(navigator, out var platformXml);
+            if (!isValidPlatform || string.IsNullOrWhiteSpace(platformXml))
```
Comment (chakradhar2) [inline src/vstest.console/TestPlatformHelpers/TestRequestManager.cs]: Can we return this condition result it self?

File: src/Microsoft.TestPlatform.ObjectModel/Constants.cs [hunk 13]
```diff
@@ -110,6 +110,8 @@ public static class Constants
 
         public const string DotNetFrameWorkStringPrefix = ".NETFramework,Version=";
 
+        public const string DotNetFramework35 = ".NETFramework,Version = v3.5";
```
Comment (harshjain2) [inline src/Microsoft.TestPlatform.ObjectModel/Constants.cs]: Is it correct to have white spaces in between? Other platform versions don't have space.

File: src/Microsoft.TestPlatform.ObjectModel/Constants.cs [hunk 14]
```diff
@@ -118,6 +120,10 @@ public static class Constants
 
         public const string DotNetFramework46 = ".NETFramework,Version=v4.6";
 
+        public const string DotNetFrameworkCore10 = ".NETCoreApp,Version=v1.0";
```
Comment (harshjain2) [inline src/Microsoft.TestPlatform.ObjectModel/Constants.cs]: What about DotNetFrameworkCore20?

File: src/Microsoft.TestPlatform.ObjectModel/Framework.cs [hunk 15]
```diff
@@ -83,5 +89,25 @@ public override string ToString()
         {
             return this.Name;
         }
+
+        public override bool Equals(Object obj)
```
Comment (harshjain2) [inline src/Microsoft.TestPlatform.ObjectModel/Framework.cs]: What is the need for this comparison?

File: src/Microsoft.TestPlatform.ObjectModel/Framework.cs [hunk 16]
```diff
@@ -53,19 +59,19 @@ public static Framework FromString(string frameworkString)
...
                     case "FrameworkCore10":
-                        frameworkName = new FrameworkName(".NETCoreApp,Version=1.0");
+                        frameworkName = new FrameworkName(Constants.DotNetFrameworkCore10);
                         break;
                     case "FrameworkUap10":
-                        frameworkName = new FrameworkName("Uap,Version=10.0");
+                        frameworkName = new FrameworkName(Constants.DotNetFrameworkUap10);
                         break;
                     default:
                         return null;
```
Comment (harshjain2) [inline src/Microsoft.TestPlatform.ObjectModel/Framework.cs]: Should we be returning the DefaultFramework here instead of null?

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs [hunk 17]
```diff
@@ -220,7 +242,7 @@ public static bool TryGetDeviceXml(XPathNavigator runSettingsNavigator, out Stri
         /// <summary>
         /// Adds node under RunConfiguration setting. Noop if node is already present.
         /// </summary>
-        private static void AddNodeIfNotPresent<T>(XPathNavigator runSettingsNavigator, string nodePath, string nodeName, T nodeValue)
+        private static void AddNodeIfNotPresent<T>(XPathNavigator runSettingsNavigator, string nodePath, string nodeName, T nodeValue, bool overwrite = false)
```
Comment (harshjain2) [inline src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: What is the need for adding `bool overwrite`?

File: src/vstest.console/CommandLine/AssemblyHelper.cs [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+            try
+            {
+                using (var assemblyStream = File.Open(filePath, FileMode.Open))
+                {
+                    frameworkName = GetFrameworkNameFromAssemblyMetadata(assemblyStream);
+                }
+            }
+            catch (Exception ex)
+            {
+                EqtTrace.Warning("GetFrameWorkFromMetadata: failed to determine TargetFrameworkVersion: {0} for assembly: {1}", ex, filePath);
```
Comment (harshjain2) [inline src/vstest.console/CommandLine/AssemblyHelper.cs]: Error?

File: src/vstest.console/CommandLine/AssemblyHelper.cs [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+            }
+
+            return frameworkName;
+        }
+
+        private Architecture MapToArchitecture(ProcessorArchitecture processorArchitecture)
+        {
+            Architecture arch = Architecture.AnyCPU;
+            // Mapping to Architecture based on https://msdn.microsoft.com/en-us/library/system.reflection.processorarchitecture(v=vs.110).aspx
+            if (processorArchitecture.Equals(ProcessorArchitecture.Amd64)
```
Comment (harshjain2) [inline src/vstest.console/CommandLine/AssemblyHelper.cs]: Ignore case?

File: src/vstest.console/CommandLine/Executor.cs [hunk 20]
```diff
@@ -195,6 +195,12 @@ private int GetArgumentProcessors(string[] args, out List<IArgumentProcessor> pr
             // Examples: processors to enable loggers that are statically configured, and to start logging,
             // should always be executed.
             var processorsToAlwaysExecute = processorFactory.GetArgumentProcessorsToAlwaysExecute();
+            foreach (var argumentProcessor in processorsToAlwaysExecute)
+            {
+                // Setup lazy initialization.
+                processorFactory.CreateArgumentProcessor(argumentProcessor.Metadata.Value.CommandName);
```
Comment (harshjain2) [inline src/vstest.console/CommandLine/Executor.cs]: Why is this required?

File: src/vstest.console/Processors/RunSpecificTestsArgumentProcessor.cs [hunk 21]
```diff
@@ -135,7 +135,7 @@ internal class RunSpecificTestsArgumentExecutor : IArgumentExecutor
         /// </summary>
         private ITestDiscoveryEventsRegistrar discoveryEventsRegistrar;
 
-#endregion
+        #endregion
```
Comment (harshjain2) [inline src/vstest.console/Processors/RunSpecificTestsArgumentProcessor.cs]: nit : remove white spaces.

File: src/Microsoft.TestPlatform.ObjectModel/Constants.cs (hunk 14 above)
Comment (codito) [reply src/Microsoft.TestPlatform.ObjectModel/Constants.cs]: We don't need it. For .net core, we interpret dynamically from the app's runtime json.

File: src/Microsoft.TestPlatform.Common/Utilities/RunSettingsProviderExtensions.cs [hunk 22]
```diff
@@ -18,7 +18,7 @@ namespace Microsoft.VisualStudio.TestPlatform.Common.Utilities
     /// </summary>
     internal static class RunSettingsProviderExtensions
     {
-        public const string EmptyRunSettings = @"<RunSettings></RunSettings>";
+        public const string EmptyRunSettings = @"<RunSettings><RunConfiguration></RunConfiguration></RunSettings>";
```
Comment (harshjain2) [inline src/Microsoft.TestPlatform.Common/Utilities/RunSettingsProviderExtensions.cs]: Why is this needed?

**Comment by smadala** [comment]
Please test Windows_NT / Release Build .

File: src/vstest.console/CommandLine/AssemblyMetadataProvider.cs [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+                EqtTrace.Info("Unable to map to Architecture, using platform: {0}", arch);
+            }
+
+            return arch;
+        }
+
+        public Architecture GetArchitectureForSource(string imagePath)
+        {
+            // For details refer to below code available on MSDN.
+            //https://code.msdn.microsoft.com/windowsapps/CSCheckExeType-aab06100#content
```
Comment (harshjain2) [inline src/vstest.console/CommandLine/AssemblyMetadataProvider.cs]: 👍

**Review by harshjain2 — APPROVED** [review]
Looks good to me from Functionality Perspective.

File: src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs (hunk 1 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs]: This data is limited and important. Keep it as Info.

File: src/Microsoft.TestPlatform.ObjectModel/Framework.cs (hunk 16 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.ObjectModel/Framework.cs]: Not change behavior in this PR.

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs (hunk 17 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: Required this to overwrite default settings.

File: src/vstest.console/CommandLine/InferHelper.cs [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+
+                        if (finalArch == null)
+                        {
+                            finalArch = arch;
+                            continue;
+                        }
+
+                        if (!finalArch.Equals(arch))
+                        {
+                            // TODO Log to console and client.
```
[the remark this answers is not in the document]
Comment (smadala) [reply src/vstest.console/CommandLine/InferHelper.cs]: Need TODOs to compact with Tpv1, We address this in other PR.

File: src/vstest.console/Processors/RunSpecificTestsArgumentProcessor.cs (hunk 21 above)
Comment (smadala) [reply src/vstest.console/Processors/RunSpecificTestsArgumentProcessor.cs]: Those are required.

**Comment by codito** [comment]
@dotnet-bot test Windows_NT / Release Build please
